Lib/_android_support.py

```diff
@@ -0,0 +1,99 @@
+import io
+import sys
+from ctypes import CDLL, c_char_p, c_int
+
+
+# The maximum length of a log message in bytes, including the level marker and
+# tag, is defined as LOGGER_ENTRY_MAX_PAYLOAD in
+# platform/system/logging/liblog/include/log/log.h. As of API level 30, messages
+# longer than this will be be truncated by logcat. This limit has already been
+# reduced at least once in the history of Android (from 4076 to 4068 between API
+# level 23 and 26), so leave some headroom.
+MAX_BYTES_PER_WRITE = 4000
+
+# UTF-8 uses a maximum of 4 bytes per character, so limiting text writes to this
+# size ensures that TextIOWrapper can always avoid exceeding MAX_BYTES_PER_WRITE.
+# However, if the actual number of bytes per character is smaller than that,
+# then TextIOWrapper may still join multiple consecutive text writes into binary
+# writes containing a larger number of characters.
+MAX_CHARS_PER_WRITE = MAX_BYTES_PER_WRITE // 4
+
+
+# When embedded in an app on current versions of Android, there's no easy way to
+# monitor the C-level stdout and stderr. The testbed comes with a .c file to
+# redirect them to the system log using a pipe, but that wouldn't be convenient
+# or appropriate for all apps. So we redirect at the Python level instead.
+def init_streams():
+    if sys.executable:
+        return  # Not embedded in an app.
+
+    # Despite its name, this function is part of the public API
+    # (https://developer.android.com/ndk/reference/group/logging).
+    # Use `getattr` to avoid private name mangling.
+    global android_log_write
+    android_log_write = getattr(CDLL("liblog.so"), "__android_log_write")
+    android_log_write.argtypes = (c_int, c_char_p, c_char_p)
```

Comment:
Did you consider creating a wrapper in C (using `PyCFunction_New`), and passing it to `init_streams`?

(I don't know if Android is already importing `ctypes` on startup elsewhere; it seems like a pretty heavy dependency for this purpose).

Reply:
Good idea – it's not already importing ctypes, so I've followed your suggestion. 

Since the initialization code no longer depends on ctypes, there's longer any valid reason for it to fail, so I've made all failures fatal.